ieeeuoft/hackathon-template
# IEEE-19 and IEEE-137: PATCH Orders

## Overview

Changed my mind about a couple of things. Both PATCH tickets corresponding to order status is done in the PR. Furthermore, instead of being done in the serializer, permission validation and status validation is done in the respective views for more clarity.

- Resolves #50 (being able to modify the status of the order) and IEEE-137 ticket.
- Created new serializer for status change method.
- Validated to ensure status change is from one of the following to another status option that is valid: 

```
For tech team

"Submitted": ["Cancelled", "Ready for Pickup"],
"Ready for Pickup": ["Picked Up"]

For reg users

"Submitted": ["Cancelled"]
```
- The above validation is done in the respective views (in the hardware and event folder respectively).
- Created 2 new views for patch capability for orders. 
- Patch ability is supported for both tech team and current users.
- There is also validation that reg users cannot change status of orders that are not from their team.

## Unit Tests Created

For tech team

- Test to ensure only allowed status changes are being made (whether checking the beginning status is allowed to change, or the end status is an acceptable final status based on initial status).
- Test to ensure login is required. 
- Test to ensure user has hardware change order capabilities.
- Test to ensure all-rules-passing success.

For regular users.

- Test to ensure only allowed status changes are being made (whether checking the beginning status is allowed to change, or the end status is an acceptable final status based on initial status).
- Test to ensure login is required. 
- Test to ensure user isn't changing orders for those who are not on their team.
- Test to ensure all-rules-passing success.

## Steps to QA

- Try creating a sample order and then testing the "orders/pk" route on swagger for both event and hardware..
- OR you could take a look at the test logic and see if it makes sense.

PS @YuYing-Liang your #320 PR was rlly helpful!
